[WIP][Qt] Add new column which uses the compatibility database to display …

scribam opened the pull request:
…the information per game

![screenshot from 2017-09-10 19-08-15](https://user-images.githubusercontent.com/25406440/30251222-b7991a78-965b-11e7-9d9f-706574d64cc7.png)

AniLeo commented:
It would probably be a good idea to have some kind of way for the user to know inside the emulator what the status means - people will see Loadable and think the game works for example.

Megamouse commented on rpcs3/rpcs3qt/game_list_frame.cpp:
```diff
@@ -925,7 +929,7 @@ void game_list_frame::resizeEvent(QResizeEvent *event)
 
 /**
- Cleans and readds entries to table widget in UI.
+ Cleans and reads entries to table widget in UI.
```
:) this was no typo. But it's fine either way

danilaml replied:
What typo? Shouldn't it stay as  `readds` as in `re-adds`?

scribam replied:
Oops, sorry about that. I reverted this change.

Megamouse commented:
maybe setText to something like: "Aquiring status" while it is in progress.
Or if no network available show: "Network unavailable"

twdarkeh commented:
Maybe color the rows, or even just the cell with the status, according to the status? So make it green for playable, yellow for ingame, etc.

rodjun commented:
Would be nice to show the RPCS3 version used on mouse over

Megamouse commented:
Please reinitiate the compatibility on our Refresh button. I just started rpcs3 with wifi disabled and after re-enabling it there was no possibility of aquiring the values

jjsat commented:
It would be good if the user could opt-out of this via option because not everyone will want the emulator to essentially transfer their game list.

AniLeo commented:
"Transfer" how?

jjsat commented:
Indirectly by accessing the compatibility database entries for all found games. Easy to extract the information even from the normal server log because I assume that the user agent from which this is accessed is identifiable. Even if not the access pattern should be pretty unique.

Megamouse commented:
we can add a dialog that asks the user once if he wants it enabled or not,
And/or add it somewhere in the settings.

AniLeo commented:
User agent is "Mozilla/5.0". No one is going to make any telemetry using the access logs, if we ever have telemetry it will be an opt-out setting. But I guess you can add it as an setting, why not.

Asinin3 commented:
Would be nice if you can see the status for different IDs because some IDs don't get updated. Possibly even right-click to go to the compat thread or search for it with the game id.

Megamouse commented:
you can already go to compat with rightclick so there is no need for that in my opinion

paulsapps commented:
What about also allowing the user to set the compatibility? Obviously this is a little more tricky to do but could be useful?

scribam commented:
Thanks for your feedback everyone.

@AniLeo I added tooltips which display the status description from https://rpcs3.net/compatibility, is it ok for you?
@Megamouse I will think about adding more information while it is in progress. I added a more descriptive message when something goes wrong with the network. Also, the refresh button should work now as you expect.
@twdarkeh I agree, some colors would be nice. I will see what I can do.
@jjsat Compatibility is now disabled by default.

Where would you like to put the checkbox to handle the compatibility activation ? Should we put it in the welcome dialog as well with a text explaining the functionality and some details about privacy  ?

rodjun commented:
Probably a better idea to just hide the compatibility tab when it's disabled. No point in showing something that just says "Disabled" on every line

raveskirza commented:
May be the compatibility column should be visible all time otherwise the users will not know it exists.
We can have the compatibility option under menu, the same way we have the option to show Game list toolbar

MarioSonic2987 commented on rpcs3/rpcs3qt/GameCompatibility.h:
```diff
@@ -0,0 +1,30 @@
+		{ "Playable", QObject::tr("Games that can be properly played from start to finish") },
+		{ "Ingame", QObject::tr("Games that go somewhere but not far enough to be considered playable") },
+		{ "Intro", QObject::tr("Games that only display some screens") },
+		{ "Loadable", QObject::tr("Games that only display some screens") },
```
Intro and Loadable have the same description. It should be **_Games that display a black screen with an active framerate_** for Loadable status.

MarioSonic2987 commented on rpcs3/rpcs3qt/GameCompatibility.h:
```diff
@@ -0,0 +1,30 @@
+		{ "Playable", QObject::tr("Games that can be properly played from start to finish") },
+		{ "Ingame", QObject::tr("Games that go somewhere but not far enough to be considered playable") },
+		{ "Intro", QObject::tr("Games that only display some screens") },
+		{ "Loadable", QObject::tr("Games that only display some screens") },
```
Intro and Loadable have the same description. It should be **_Games that display a black screen with an active framerate_** for Loadable status.

scribam replied:
Fixed, thanks!

Renusek commented:
IMO it should be a separate column, but maybe colored game title (maybe you could disable that in options if you want all game titles to be black.

Megamouse commented:
Just add some smileys as icons. no need for ugly colors

twdarkeh commented:
As for where to put the checkbox to enable it, settings -> emulator -> UI settings, below "show welcome screen" would be the best place.

digitaldude555 commented:
Just have playable and not playable. Nothing/Loadable/Intro/Ingame is not relevant for users.

Asinin3 commented:
@digitaldude555 I don't see why not, they can open rpcs3 and see if there games have been making any progress. Also very useful for testers especially if it can show other IDs for the same game and the date it was tested on.

rodjun commented:
@digitaldude555 Why have anything at all if people can just launch the game and find out that status

scribam commented:
Colors and date added:

![screenshot from 2017-09-11 23-30-55](https://user-images.githubusercontent.com/25406440/30298275-1b8dab3a-974b-11e7-8760-807e3c171515.png)

greentop commented:
-Ubuntu 16.04 
(Compatibility manually enabled for testing.)

![rpcs3 ubuntu 16 04 pr3426](https://user-images.githubusercontent.com/2325766/30299769-40ee0a7c-971f-11e7-9277-28d1597b58e9.png)

Luffykun007 commented:
@scribam This build says that Qt5Network.dll is missing from my computer and does not work for me.

danilaml commented:
>This build says that Qt5Network.dll is missing from my computer and does not work for me.

A proper illustration of why I think that https://github.com/RPCS3/rpcs3/pull/3432 is a better solution to the existing one.

Asinin3 commented:
@scribam Would it be possible to sort by last updated? Or maybe even status?

raveskirza commented:
https://ci.appveyor.com/project/rpcs3/rpcs3/build/0.0.3-6154/artifacts
I get this error when I use the size scroll bar. Looks like the game list is refreshed for each step moved in the scroll bar. If you disable the compatibility, RPCS3 doesn't crash

![image](https://user-images.githubusercontent.com/28158623/30405156-4a402d22-98b8-11e7-9682-2b729a82ed6c.png)

Megamouse commented:
yeah you should only refresh compatibility on certain refreshs, like the refresh button or first time loading.

scribam commented:
@Luffykun007 Can you try again to run it under Windows ? I did a rebase which include #3432 and I hope it will fix the issue about Qt5Network.dll.

Also, there is now a checkbox to enable/disable the compatibility feature in the Settings window, Emulator tab.

AniLeo commented:
Maybe we could have some kind of datetime comparing mechanism, if the game was tested after the build's datetime then warn the user that it may not work as they have an older build? 

On that note, having the build's compilation date on program title could be useful too so we don't have to check when is it from by looking for commit information.

Asinin3 commented:
@scribam I had the same issue as Luffy before, and I still have it with your new build.

Megamouse commented:
@Asinin3 you of course need to copy over the new files

rodjun commented:
I still don't really like the idea of the colored circles, feels unnatural next to other UI elements. An option to disable just the circles would be appreciated even if it's just a hidden ini config without fancy UI checkboxes

Luffykun007 commented:
@Scribam works for me now but I am having this error creating SSL context when it will check the compatibility and the emulator crashes.

![capturar](https://user-images.githubusercontent.com/26087621/30710119-e2c519ae-9eda-11e7-9cf4-6287b4e4f823.JPG)

Megamouse commented:
@Luffykun007 
This seems to be a Qt bug. But some say you have to copy some ssl dlls to the rpcs3 folder

danilaml commented on rpcs3/rpcs3qt/game_compatibility.h:
```diff
@@ -0,0 +1,59 @@
+	QString m_color;
+};
+
+const std::map<QString, game_compatibility_status> GAME_COMPATIBILITY_STATUS_DATA = {
```
Why make this global? It's only ever used in `cpp`. Could be moved there as well (maybe even in a anonymous namespace).

scribam replied:
Moved to the cpp file, thanks.

danilaml commented on rpcs3/rpcs3qt/game_compatibility.h:
```diff
@@ -0,0 +1,59 @@
+	{
+		fill(Qt::transparent);
+
+		m_painter = std::make_unique<QPainter>(this);
```
Does it really need a painter as a member? It's used only once here. Could be just a local var.

scribam replied:
Not really indeed. Use a local var now, thanks.

danilaml commented:
Works for me with the exe from the latest appveyor build. No copying was required.

Megamouse commented:
@danilaml that might be because you have Qt installed and in path

danilaml commented:
@Megamouse yeah, you are correct. What bug though? As I understand, if you use SSL it should be provided by your system/yourself, since it's may be illegal to redistribute cryptography libs such as OpenSSL in some parts of the world.

Megamouse commented:
@danilaml yeah I mistook the issue for open. It is already closed though. sorry. 
But the common "fix" is indeed to provide the SSL stuff yourself.

@scribam what did you change?
I compiled some stuff for you in discord btw

scribam commented:
@Megamouse I did a rebase on master and took care of few things that danilaml reviewed.

Asinin3 commented:
Sorry, I didn't copy the dll over my fault. Anyway Now I'm getting this after enabling the feature and hitting refresh. 
![image](https://user-images.githubusercontent.com/15137035/30722918-e9fa04b2-9f72-11e7-847c-0727ab714be4.png)

MelchiorGaspar commented:
The new column is not even displaying for me or even on the drop-down menu list... :( :/

AniLeo commented:
It's better to use the API than a local database so we can get newer results in real time. Newer results will encourage people to update their builds if a warning is included when the build tested is more recent.

Maybe the results could be somewhat cached? Could be useful for people that have a lot of games perhaps, but then again it would have to ping the API again to check for updates.

MelchiorGaspar commented:
cached would be good....  just add a context menu entry or a menu entry some where to refresh the cache..
O_o

scribam commented:
This PR needs more work than I expected and I do not think I am skilled enough to find a proper way to integrate with the current code. Maybe I will work on it again, later, and submit a more mature PR. Until this time, feel free to use my code as you want to bring this feature to RPCS3 :grin:
